modules/dnn/src/tensorflow/tf_importer.cpp

```diff
@@ -1110,23 +1113,28 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
                 else
                 {
                     inpLayout = DATA_LAYOUT_NHWC;
+                    changedType = newShapeSize == 4 && !hasSwap;
                 }
             }
         }
         layerParams.set("dim", DictValue::arrayInt<int*>(newShape.ptr<int>(), newShapeSize));
 
-        int id = dstNet.addLayer(name, "Reshape", layerParams);
-        layer_id[name] = id;
+        std::string setName = changedType ? name + "/realReshape" : name;
+
+        int id = dstNet.addLayer(setName, "Reshape", layerParams);
+        layer_id[setName] = id;
 
         // one input only
         connect(layer_id, dstNet, inpId, id, 0);
-        inpId = Pin(name);
+        inpId = Pin(setName);
 
         if ((inpLayout == DATA_LAYOUT_NHWC || inpLayout == DATA_LAYOUT_UNKNOWN || inpLayout == DATA_LAYOUT_PLANAR) &&
             newShapeSize == 4 && !hasSwap)
         {
             int order[] = {0, 3, 1, 2};  // Transform back to OpenCV's NCHW.
-            addPermuteLayer(order, name + "/nchw", inpId);
+
+            setName = changedType ? name : name + "/nchw";
+            addPermuteLayer(order, setName, inpId);
```

Comment:
***For reviewers***:
This issue is triggered when this if condition is entered. `addPermuteLayer` sets the output name to `name + "/nchw"`, leading to the incorrect input fetched by the next layer.

Reply:
Why do we need `name + "/nchw"` node at the end? I don't think other layers know this node by name.

Reply:
@rogday It was introduced by you in a pull request a year ago: https://github.com/opencv/opencv/pull/20190/files#diff-f5872c76075aa1f06f357d1dc27448bf1a02a28b79e62d92e8b7aa47311b57dfR1007

Reply:
@fengyuentau, it's a refactoring PR that broke history a little bit.

Reply:
Heh, I was investigating it [already](https://github.com/opencv/opencv/pull/19641#issuecomment-892756609), but didn't get the answer.

Reply:
@rogday I've investigated this. So, there's a sample in tests, where the next condition is not met:
https://github.com/opencv/opencv/pull/22448/files#diff-f5872c76075aa1f06f357d1dc27448bf1a02a28b79e62d92e8b7aa47311b57dfR1105

It's not met cause newShape has values: [1, 2, 4, 3].
So, there will not be permutation before the actual Reshape node.
The test sample is: reshape_layer with the dnn/tensorflow/reshape_layer_net.pb graph. Which has only input and Reshape layer(look at the screenshot).
![Screenshot from 2022-09-06 19-20-01](https://user-images.githubusercontent.com/22273578/188699917-b3eba7a1-c0dd-40a5-86db-d0c0ab7b5960.png)

The reference output has shape: output shape: (1, 3, 2, 4).
So, the loaded graph in opencv will look like:
input->Reshape->Reshape/nchw.

Also this behaviour can be observed in not_implemented_layer_net.pb graph, which has 2 inputs, 1 is going to not_implemented_layer, the 2nd is going to reshape, after that outputs from this layers are multiplied.

So, the problem, that you've covered may be defined like:
1). Input layer of reshape has NHWC layout.
2). By some reason(the question here is why don't we reshape, if condition newShape.at<int>(1) == 1 is not met) the permutation from HNWC to NCHW layouts is not performed.
3). Next Reshape is performed only with NCHW data layout.
4). Data is always permuted from NCHW to NHWC data format at the end if the source of Reshape in original graph is NHWC/ Even if it wasn't permuted from NHWC to NCHW at input.

I don't see any reasons, why there's no permutation at the start. Cause usually when you have graph, where all layers have 1 data layout.
I can investigate it lately, but this is not what this PR is about. Also the changing of naming the original node name to last permutation layer breaks several tests.